Adding NST with AdaIN

ariG23498 opened the pull request:
In this PR we try to implement Neural Style Transfer with [Adaptive Instance Normalization](https://arxiv.org/abs/1703.06868).

[Colab Notebook](https://colab.research.google.com/drive/1mn9WrC6eT3hCVlVBGoILQScsP5krQibV?usp=sharing)
[Hugging Face demo](https://huggingface.co/spaces/ariG23498/nst)

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+        self.decoder = decoder
+        self.ada_in = ada_in
+        self.loss_net = loss_net
+        self.lamb = lamb
```
if this is lambda, let's just call it `lambda`.  The lamb abbreviation can cause confusion.

ritwikraha replied:
We think ```style_weight``` is more accurate here, so we replaced ```lamb``` with that.

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+
+        with tf.GradientTape() as tape:
+            # Encode the style and content image.
+            style_enc = self.encoder(style)
```
enc -> encoded

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+            content_enc = self.encoder(content)
+
+            # Compute the AdaIN feature map.
+            t = self.ada_in(style=style_enc, content=content_enc)
```
same here, enc -> encoded.

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+            content_enc = self.encoder(content)
+
+            # Compute the AdaIN feature map.
+            t = self.ada_in(style=style_enc, content=content_enc)
```
anything more descriptive we can call `t`?  it's ok if not.

ritwikraha replied:
The term ```t``` (target feature maps) is taken from the paper itself
![image](https://user-images.githubusercontent.com/44690292/141061422-b5eb32b2-bff3-4fcd-9300-3a2dc08416d2.png)
So we think its best to leave it as ```t```

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+            recons_image = self.decoder(t)
+
+            # Compute the losses.
+            recons_vgg_feats = self.loss_net(recons_image)
```
recons->reconstructed

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+            t = self.ada_in(style=style_enc, content=content_enc)
+
+            # Generate the neural style transferred image.
+            recons_image = self.decoder(t)
```
recons->reconstructed all over

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+
+            # Compute the losses.
+            recons_vgg_feats = self.loss_net(recons_image)
+            style_vgg_feats = self.loss_net(style)
```
feats -> features

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+            recons_vgg_feats = self.loss_net(recons_image)
+            style_vgg_feats = self.loss_net(style)
+            loss_content = self.loss_fn(t, recons_vgg_feats[-1])
+            for inp, out in zip(style_vgg_feats, recons_vgg_feats):
```
feats -> features

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+            style_vgg_feats = self.loss_net(style)
+            loss_content = self.loss_fn(t, recons_vgg_feats[-1])
+            for inp, out in zip(style_vgg_feats, recons_vgg_feats):
+                mean_inp, std_inp = get_mean_std(inp)
```
maybe inp => features?

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+        style, content = inputs
+        loss_content = 0.0
+        loss_style = 0.0
+        style_enc = self.encoder(style)
```
same comments here as train_step

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+This callback is used to visualize the style transfer capability of
+the model on each epoch. The objective of style transfer is not
+quantified properly and is mostly upto the audience to call is good
+or bad. For this reason, visualization falls into a key aspect of
```
-> change to:

visualization is a key aspect of training the model

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+class TrainMonitor(tf.keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        # Encode the style and content image.
+        test_style_enc = self.model.encoder(test_style)
```
same renames, encoded

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+"""
+# Conclusion
+
+Adaptive Instance Normalization is an important step in the direction
```
we can remove "is an important step in the direction of Neural Style Transfer research because it" resulting in...

`Adaptive Instance Normalization allows arbitrary style
transfer in real time`

LukeWood commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+Adaptive Instance Normalization is an important step in the direction
+of Neural Style Transfer research beacuse it allows arbitrary style
+transfer in real time. It is also important to note that the novel
+proposition of the authors is to achive this only by aligning the
```
achive -> achieve

LukeWood reviewed (changes requested):
Thank you for your contribution.  This is a really great example, I left some comments.  Please address them and reply back.

I really love real time Neural Style Transfer (I published https://imstyle.app awhile back...) so it's great to see some extra examples using it.

LukeWood commented:
Also note you will need to format your file after the changes to get the CI workflow to pass.

```
black examples/vision/adain.py
```
should work

ritwikraha reviewed (commented):
@LukeWood thanks for the detailed review. We have fixed most of the things you mentioned and will push the changes for review shortly

ariG23498 commented on examples/vision/adain.py:
```diff
@@ -0,0 +1,675 @@
+"""
```
Thank you for your kind words!

ariG23498 commented:
@LukeWood Thanks for your thorough review. We have taken care of each of the suggestions. After the changes we have used the black formatting yet the CI/CD pipeline does not seem to work for us. Could you help here?

LukeWood commented:
Looks good!  Please run:
```
cd scripts
python3 autogen.py add_example vision/adain.py
```
to add the generated files

ariG23498 commented:
@LukeWood we have added the generated files.

LukeWood commented:
@ariG23498 are the images in examples/vision/img/adain/adain_25_0.png correct?  I see a deeply red image for every style/every content image.  Did something go wrong in the figure generation?

LukeWood reviewed (changes requested):
Let's make sure those figures are correct.  If needed you can fit the model for longer locally and host the resulting diagrams on IMGUR.

LukeWood commented:
Ah -- it's because we only ran for 1 epoch.  I wonder.... do you think it's more informative/impressive to link to results from 30 epochs?

ariG23498 commented:
Hey @LukeWood thank you for observing this.

In the introduction, we have added the output for a 30 epoch trained model. The image is 

![img](https://i.imgur.com/zDjDuea.png)

We have also added a hugging face demo for style transfer with the pretrained models. The link is: https://huggingface.co/spaces/ariG23498/nst

I think this should be enough for the reader to familiarise with the output of a trained mode. What do you think?

LukeWood commented:
Ah fantastic - thank you.

Would you mind running `python3 autogen.py add_example vision/adain.py`
 one more time?  I updated one more link :)

Then I think we are good!

ariG23498 commented:
Hey @LukeWood we did not generate the files again. Went in to the md and ipynb and made the necessary updates. Hope this is alright 😄

LukeWood commented:
works for me - thanks!!!

LukeWood commented:
@fchollet for a last review

fchollet reviewed (commented):
Thanks for the PR -- this is a very cool example! 👍  Thanks @LukeWood for the review too!

I pushed some copyedits, please pull them an regenerate the files. Then we can merge.

ariG23498 commented:
@LukeWood @fchollet I have pulled the copy edits and generated the files again. Thanks for the review 😄

fchollet commented:
Thanks for the great contribution!
